cub/cub/warp/specializations/warp_exchange_shfl.cuh

```diff
@@ -0,0 +1,327 @@
+/******************************************************************************
+ * Copyright (c) 2011-2023, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include <cub/config.cuh>
+#include <cub/util_ptx.cuh>
+#include <cub/util_type.cuh>
+
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+
+template <typename InputT,
+          int ITEMS_PER_THREAD,
+          int LOGICAL_WARP_THREADS = CUB_PTX_WARP_THREADS>
+class WarpExchangeShfl
+{
+  static_assert(PowerOfTwo<LOGICAL_WARP_THREADS>::VALUE,
+                "LOGICAL_WARP_THREADS must be a power of two");
+
+  static_assert(ITEMS_PER_THREAD == LOGICAL_WARP_THREADS,
+                "WARP_EXCHANGE_SHUFFLE currently only works when ITEMS_PER_THREAD == "
+                "LOGICAL_WARP_THREADS");
+
+  constexpr static bool IS_ARCH_WARP = LOGICAL_WARP_THREADS == CUB_WARP_THREADS(0);
+
+  template <typename OutputT, int IDX>
+  class CompileTimeArrayImpl; // fwd decl for partial specializations
+
+  template <typename OutputT> // terminating partial specialization
+  class CompileTimeArrayImpl<OutputT, ITEMS_PER_THREAD>
+  {
+  protected:
+    // used for dumping back the individual values after transposing
+    InputT (&output_items)[ITEMS_PER_THREAD];
```

Comment:
I am wondering, would it be possible to use `cuda::std::span<InputT, ITEMS_PER_THREAD>` without nvcc messing with registers?

Reply:
@senior-zero what is your preference here? I looked at the `cuda::std::span` implementation and saw that it is just the usual ISO C++ implementation (pointer + size for dynamic extent). My only worry here is the pointer-aliasing/escape analysis. In general references are slightly easier to optimise (since they can't be reseated) and I'm always worried `nvcc` cannot optimize out the transient pointer and then the whole lot gets dropped into local memory and it's game over.

Reply:
@pb-dseifert you could leave experimenting with it to the follow-up PR if you'd like. I don't think it's a blocker for merging this PR.

Reply:
Yeah as I said, this is more me thinking out loud. No change requested